wagn/rails/controllers/card_controller.rb

```diff
@@ -38,93 +32,126 @@ def delete
+
+  def setup
     request.format = :html if !params[:format] # is this used??
     Card::Cache.renew
     Card::Env.reset controller: self
-    Card::Auth.set_current_from_session
+  end
 
-    if params[:id] && !params[:id].valid_encoding?
-      # slightly better way to handle encoding issues (than the rescue in
-      # load_id)
-      # we should find the place where we produce these bad urls
-      params[:id] = params[:id].force_encoding('ISO-8859-1').encode('UTF-8')
+  def authenticate
+    if params[:token]
+      Card::Auth.set_current_from_token params[:token], params[:user_id]
+    else
+      Card::Auth.set_current_from_session
     end
   end
 
   def load_id
-    params[:id] ||=
-      case
-      when Card::Auth.needs_setup? && Card::Env.html?
-        params[:card] = { type_id: Card.default_accounted_type_id }
-        params[:view] = 'setup'
-        ''
-      when params[:card] && params[:card][:name]
-        params[:card][:name]
-      when Card::Format.tagged(params[:view], :unknown_ok)
-        ''
-      else
-        Card.setting(:home) || 'Home'
+    params[:id] =
+      case params[:id]
+      when '*previous' then return card_redirect(Card::Env.previous_location)
+      when nil         then find_id
+      else                  validate_id_encoding params[:id]
       end
-  rescue ArgumentError # less than perfect way to handle encoding issues.
-    raise Wagn::BadAddress
   end
 
   def load_card
-    if params[:id] == '*previous'
-      return card_redirect(Card::Env.previous_location)
-    end
-
-    opts = card_attr_from_params
-    @card =
-      if params[:action] == 'create'
-        # FIXME: we currently need a "new" card to catch duplicates (otherwise
-        # save will just act like a normal update)
-        # I think we may need to create a "#create" instance method that
-        # handles this checking.
-        # that would let us get rid of this...
-        Card.new opts
-      else
-        mark = params[:id] || opts[:name]
-        Card.fetch mark, new: opts
-      end
+    @card = get_main_card
     raise Card::NotFound unless @card
 
-    @card.select_action_by_params params
+    @card.select_action_by_params params #
     Card::Env[:main_name] = params[:main] || (card && card.name) || ''
 
-    render_errors if card.errors.any?
-    true
+    card.errors.any? ? render_errors : true
   end
 
   def refresh_card
     @card = card.refresh
   end
 
-  protected
+  # ----------( HELPER METHODS ) -------------
+
+  def get_main_card
+    opts = card_attr_from_params
+    if params[:action] == 'create'
+      # FIXME: we currently need a "new" card to catch duplicates
+      # (otherwise save will just act like a normal update)
+      # We may need a "#create" instance method to handle this checking?
+      Card.new opts
+    else
+      mark = params[:id] || opts[:name]
+      Card.fetch mark, new: opts
+    end
+  end
+
+  def card_attr_from_params
```

Comment:
Assignment Branch Condition size for card_attr_from_params is too high. [15.43/15]